### DISCUSSION TITLE
Ask and answer. Report your problems or feedback

### DISCUSSION THREAD

**Discussion by TemaSM** [RealSpeaker/telegraf-session-local#113] [Q&A]
Ask questions about this lib, report your problems or just give us feedback

**Comment by niusia-ua** [comment]
Hello. Am I correct in understanding that the asynchronous adapter (for example, `LocalSession.storageFileAsync`) is not only created asynchronously, that you have to wait for it before starting the bot, but also performs read/write operations asynchronously? I ask because in the examples with the asynchronous adapter, only the creation is asynchronous, but not the operations.

**Reply by TemaSM** [reply]
Hi, Nazar. Thanks for your question!

`storageFileAsync` is adapter of [library called `lowdb` ](https://github.com/typicode/lowdb/tree/v1.0.0) (we're using it in `telegraf-session-local`)
Under the hood it uses basic node's `fs` module to operate with filesystem (read/write), here's source of `storageFileAsync`:
https://github.com/typicode/lowdb/blob/v1.0.0/src/adapters/FileAsync.js

So, you can access `lowdb` instance on Telegraf's context like this:
```js
bot.on('text', (ctx, next) => {
  // `property` (from LocalSession constructor) + 'DB' is a name of ctx property which contains lowdb instance, default = `sessionDB`
  console.log(ctx.sessionDB)

  return next()
})
```
and make a direct [call to any of `lowdb` methods](https://github.com/typicode/lowdb/tree/v1.0.0#how-to-query) (which are basically methods of `lodash` lib), including read/write:
```js
// id - identifier of session (https://github.com/RealSpeaker/telegraf-session-local/blob/master/lib/session.js#L36)
ctx.sessionDB.get('sessions').push({ id: '123:123', data: { hello: 'world' } }).write()
  .then(() => console.log('Added session for user 123 and saved DB'))
// write/read returns Promise if `lowdb` instance initialized by using *Async adapter
```
In basic usage scenario of `telegraf-session-local` you don't even need to worry about sync/async operations, but keep in mind that sync operations are I/O blocking which slows down Nodejs.

---
Moreover, `lowdb` instance gives an access to built in `lodash` API, like this:
```js
ctx.sessionDB._
```
\- by using it you can extend `lowdb` instance (and LocalSession instance itself of course) as you want, like this:
```js
const localSession = new LocalSession()
localSession.DB._.mixin({
  first: (arr) => {
    return arr[0]
  }
})

// later somewhere access your mixin `first()` directly via `LocalSession` instance:
localSession.DB.get('posts').first().value()
// or via Telegraf's context:
bot.command('post', (ctx, next) => {
  const welcome = ctx.sessionDB.get('posts').first().value()
  return ctx.reply(JSON.stringify(welcome))
})
```

And you can iterate over all sessions, like this:
```js
const sessions = localSession.get('sessions').valueOf()
sessions.forEach((session) => {
  if (session.data.notificationsEnabled) {
    const id = session.id.split(':')[0]
    bot.telegram.sendMessage(id, 'Yo! New notification')
  }
})
```

Hope all this info will help you to build new awesome bot for Telegram 👍

**Reply by niusia-ua** [reply]
Thanks for such a quick and informative reply! It will definitely be useful for me.
